app/helpers/ExerciseConfig/Validation/ExerciseConfigValidator.php

```diff
@@ -59,49 +70,78 @@ private function checkEnvironments(ExerciseConfig $config, array $variablesTable
    * @throws ExerciseConfigException
    */
   private function checkPipelines(ExerciseConfig $config) {
-
-    // define reusable check function
-    $checkFunc = function (array $pipelines) {
-      foreach ($pipelines as $pipeline => $pipelineVars) {
-        try {
-          $entity = $this->pipelines->findOrThrow($pipeline);
-        } catch (NotFoundException $e) {
-          throw new ExerciseConfigException("Pipeline '$pipeline' not found");
-        }
-        $this->checkPipelineVariables($entity, $pipelineVars);
-      }
-    };
-
     foreach ($config->getTests() as $test) {
       // check default pipelines
-      $checkFunc($test->getPipelines());
+      $this->checkPipelinesSection($test->getPipelines());
 
       // go through all environments in test
       foreach ($test->getEnvironments() as $envId => $environment) {
         // check pipelines in environment
-        $checkFunc($environment->getPipelines());
+        $this->checkPipelinesSection($environment->getPipelines(), $envId);
       }
     }
   }
 
   /**
-   * @param Pipeline $pipelineEntity
-   * @param PipelineVars $pipelineVars
+   * @param array $pipelines
+   * @param string $environment
+   * @throws ExerciseConfigException
    */
-  private function checkPipelineVariables(Pipeline $pipelineEntity, PipelineVars $pipelineVars) {
-    $pipeline = $this->loader->loadPipeline($pipelineEntity->getPipelineConfig()->getParsedPipeline());
-    // @todo
+  private function checkPipelinesSection(array $pipelines, ?string $environment = NULL) {
+    /**
+     * @var string $pipelineId
+     * @var PipelineVars $pipelineVars
+     */
+    foreach ($pipelines as $pipelineId => $pipelineVars) {
+      $pipelineEntity = $this->pipelines->get($pipelineId);
+      if ($pipelineEntity === NULL) {
+        throw new ExerciseConfigException("Pipeline '$pipelineId' not found");
+      }
+
+      $pipeline = $this->loader->loadPipeline($pipelineEntity->getPipelineConfig()->getParsedPipeline());
+      $dataInBoxes = $pipeline->getDataInBoxes();
+      $inBoxNames = array_map(function (Box $box) { return $box->getName(); }, $dataInBoxes);
+      $variables = $pipelineVars->getVariablesTable();
+      $variableNames = array_map(function (Variable $box) { return $box->getName(); }, $variables->getAll());
+
+      if ($variableNames !== array_unique($variableNames)) {
```

Comment:
I believe this check is redundant, have a look at VariablesTable class... Also tiny detail: line 105 name of the variable $box is a bit misleading